runtime/warp.cu

```diff
@@ -80,4 +80,164 @@ __device__ void warpReduceTIDX(
   }
 }
 
+// specialized version with T = std::complex<float>
+template <
+    bool SINGLE_WARP,
+    typename Func,
+    typename _dim3ti,
+    typename _dim3bd>
+__device__ void warpReduceTIDX(
+    std::complex<float>& out,
+    const std::complex<float>& inp_val,
+    Func reduction_op,
+    const _dim3ti& thread_idx,
+    const _dim3bd& block_dim,
+    std::complex<float>* shared_mem,
+    bool read_write_pred,
+    std::complex<float> init_val) {
+  constexpr int WARP_SIZE = 32;
+
+  // Assume input padded to multiples of a warp
+  std::complex<float> reduce_val = init_val;
+
+  // Do warp reduction
+  if (read_write_pred) {
+    reduce_val = inp_val;
+  }
+
+  // Reduce within each warp
+  for (int i = 16; i >= 1; i /= 2) {
+    float real = __shfl_xor_sync(0xffffffff, reduce_val.real(), i, WARP_SIZE);
+    float imag = __shfl_xor_sync(0xffffffff, reduce_val.imag(), i, WARP_SIZE);
```

Comment:
Looks like `__shfl_xor_sync` is the only change? If so, could you instead of overload `__shfl_xor_sync` and keep `warpReduceTIDX` untouched?

Reply:
great idea! done.